metadata-ingestion/src/datahub/utilities/file_backed_collections.py

```diff
@@ -1,87 +1,162 @@
 import collections
+import pathlib
 import sqlite3
-import tempfile
 from dataclasses import dataclass, field
 from typing import (
     Any,
     Callable,
+    Dict,
     Generic,
     Iterator,
     List,
     MutableMapping,
+    Optional,
     OrderedDict,
     Sequence,
     Tuple,
     TypeVar,
     Union,
 )
 
+_DEFAULT_TABLE_NAME = "data"
+_DEFAULT_MEMORY_CACHE_MAX_SIZE = 2000
+_DEFAULT_MEMORY_CACHE_EVICTION_BATCH_SIZE = 200
+
 # https://docs.python.org/3/library/sqlite3.html#sqlite-and-python-types
 SqliteValue = Union[int, float, str, bytes, None]
 
 _VT = TypeVar("_VT")
 
 
+@dataclass
+class _SqliteConnectionCache:
+    """
+    If you pass the same filename to multiple FileBacked* objects, they will
+    share the same underlying database connection. This also does ref counting
+    to drop the connection when appropriate.
+    """
+
+    _ref_count: Dict[pathlib.Path, int] = field(default_factory=dict)
+    _sqlite_connection_cache: Dict[pathlib.Path, sqlite3.Connection] = field(
+        default_factory=dict
+    )
+
+    def get_connection(self, filename: pathlib.Path) -> sqlite3.Connection:
+        if filename not in self._ref_count:
+            conn = sqlite3.connect(filename, isolation_level=None)
+
+            # These settings are optimized for performance.
+            # See https://www.sqlite.org/pragma.html for more information.
+            # Because we're only using these dbs to offload data from memory, we don't need
+            # to worry about data integrity too much.
+            conn.execute('PRAGMA locking_mode = "EXCLUSIVE"')
+            conn.execute('PRAGMA synchronous = "OFF"')
+            conn.execute('PRAGMA journal_mode = "MEMORY"')
+            conn.execute(f"PRAGMA journal_size_limit = {100 * 1024 * 1024}")  # 100MB
+
+            self._ref_count[filename] = 0
+            self._sqlite_connection_cache[filename] = conn
+
+        self._ref_count[filename] += 1
+        return self._sqlite_connection_cache[filename]
+
+    def drop_connection(self, filename: pathlib.Path) -> None:
+        self._ref_count[filename] -= 1
```

Comment:
what happens if we mistakenly call drop_connection on a filename that does not have any open refs?